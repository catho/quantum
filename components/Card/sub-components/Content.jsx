import styled from 'styled-components';
import PropTypes from 'prop-types';

import {
  spacing,
  baseFontSize as defaultBaseFontSize,
} from '../../shared/theme';

const Content = styled.div`
  margin: 0;

  ${({
    theme: {
      baseFontSize,
      spacing: { small, medium },
    },
  }) => `
    padding: ${small}px ${medium}px;
    font-size: ${baseFontSize * 0.875}px;
  `}
`;

Content.propTypes = {
  theme: PropTypes.shape({
    baseFontSize: PropTypes.number,
    spacing: PropTypes.object,
  }),
};

Content.defaultProps = {
  theme: {
    spacing,
    baseFontSize: defaultBaseFontSize,
  },
};

Content.displayName = 'Card.Content';

export default Content;
