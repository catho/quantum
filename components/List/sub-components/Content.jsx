import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import Header from './Header';
import SubHeader from './SubHeader';

const ContentBox = styled.div`
  display: flex;
  flex-direction: column;
`;

ContentBox.displayName = 'Content';

class Content extends React.Component {
  _renderContent = content => {
    const { header: propsHeader, subheader: propsSubHeader } = this.props;

    if (typeof content === 'string') {
      return <Header>{content}</Header>;
    }

    const { header = propsHeader, subheader = propsSubHeader } = content;

    return (
      <>
        <Header>{header}</Header>

        {subheader && <SubHeader>{subheader}</SubHeader>}
      </>
    );
  };

  render() {
    const { content, children } = this.props;

    return <ContentBox>{children || this._renderContent(content)}</ContentBox>;
  }
}

Content.defaultProps = {
  subheader: '',
  header: '',
  content: {},
  children: null,
};

Content.propTypes = {
  header: PropTypes.string,
  subheader: PropTypes.string,
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      header: PropTypes.string,
      subheader: PropTypes.string,
    }),
  ]),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

Content.displayName = 'List.Content';

export default Content;
