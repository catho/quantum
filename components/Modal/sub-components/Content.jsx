import styled from 'styled-components';
import PropTypes from 'prop-types';
import Card from '../../Card';
import { spacing } from '../../shared/theme';

const Content = styled(Card.Content)`
  font-size: 16px;
  max-height: 70vh;
  overflow-y: auto;
  padding: ${({
    theme: {
      spacing: { medium },
    },
  }) => `${medium}px`};
`;

Content.propTypes = {
  theme: PropTypes.shape({
    spacing: PropTypes.object,
  }),
};

Content.defaultProps = {
  theme: { spacing },
};

export default Content;
