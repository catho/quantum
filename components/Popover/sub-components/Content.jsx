import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import { components, spacing, colors, breakpoints } from '../../shared/theme';
import getArrow from '../arrowProperties';
import Button from '../../Button';
import { shadow } from '../../shared';

const _getColors = ({
  skin,
  theme: {
    components: {
      popover: {
        skins: {
          [skin]: { text, background },
        },
      },
    },
  },
  inverted,
}) => `
  background-color: ${inverted ? text : background}
  color: ${inverted ? background : text}
`;

const _getTextColor = ({
  skin,
  theme: {
    components: {
      popover: {
        skins: {
          [skin]: { background, text },
        },
      },
    },
  },
  inverted,
}) => `color: ${inverted ? background : text};`;

const PopoverContent = styled.div`
  ${shadow(5)};
  align-items: start;
  display: flex;
  border-radius: 4px;
  font-size: 16px;
  ${({
    theme: {
      spacing: { xsmall },
    },
  }) => `padding: ${xsmall}px;`}

  position: absolute;
  line-height: 0;
  transition: opacity 0.2s ease-in-out, visibility 0.2s ease-in-out;
  z-index: 100;

  &:before {
    ${getArrow};
  }

  ${_getColors}
`;

const CloseButton = styled(Button.Icon).attrs({
  icon: 'close',
})`
  ${_getTextColor}
  display: inherit;
  height: auto;
  ${({
    theme: {
      spacing: { medium },
    },
  }) => `margin-left: ${medium}px;`}

  opacity: 0.8;
  padding: 0;
  transition: opacity 0.4s ease;
  width: auto;

  :hover {
    background: none;
    opacity: 1;
    ${_getTextColor}
  }
`;

const PopoverChildren = styled.div``;

class Content extends Component {
  constructor(props) {
    super(props);

    this.wrapper = document.createElement('section');
  }

  componentDidMount() {
    document.body.appendChild(this.wrapper);
  }

  componentWillUnmount() {
    document.body.removeChild(this.wrapper);
  }

  render() {
    const {
      placement,
      children,
      onPopoverClose,
      theme,
      skin,
      inverted,
      ...rest
    } = this.props;

    return ReactDOM.createPortal(
      <PopoverContent
        theme={theme}
        inverted={inverted}
        placement={placement}
        skin={skin}
        ref={element => {
          this.innerContentRef = element;
        }}
        {...rest}
      >
        <PopoverChildren>{children}</PopoverChildren>
        <CloseButton
          skin={skin}
          theme={theme}
          inverted={inverted}
          onClick={onPopoverClose}
        />
      </PopoverContent>,
      this.wrapper,
    );
  }
}

CloseButton.displayName = 'CloseButton';
PopoverChildren.displayName = 'PopoverChildren';

Content.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  inverted: PropTypes.bool,
  placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  onPopoverClose: PropTypes.func,
  skin: PropTypes.oneOf(['neutral', 'success', 'warning', 'error']),
  theme: PropTypes.shape({
    colors: PropTypes.object,
    spacing: PropTypes.object,
    breakpoints: PropTypes.array,
    components: PropTypes.shape({
      popover: PropTypes.object,
      button: PropTypes.object,
    }),
  }),
};

Content.defaultProps = {
  inverted: false,
  placement: 'top',
  onPopoverClose: () => {},
  skin: 'neutral',
  theme: {
    colors,
    spacing,
    breakpoints,
    components: {
      popover: components.popover,
      button: components.button,
    },
  },
};

export default Content;
