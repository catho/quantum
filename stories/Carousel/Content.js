const content = [
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '1 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '2 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '3 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '4 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '5 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '6 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '7 Lorem ipsum dolor avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'carousel title',
    description: '8 Lorem ipsum dolor avec celestium of soterum',
  },
];

const contentWithLongTexts = [
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'Image description',
    title: 'This is a long title for carousel',
    description:
      '1 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '2 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '3 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '4 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '5 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '6 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '7 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
  {
    imagePath: 'https://assets.catho.com.br/logo/svg/blue.svg',
    imageDescription: 'a11y image description',
    title: 'This is a long title for carousel',
    description:
      '8 Lorem ipsum dolor avec celestium of soterum avec celestium of soterum',
  },
];

export { content, contentWithLongTexts };
