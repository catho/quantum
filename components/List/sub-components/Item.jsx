import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import Icon from '../../Icon';
import Content from './Content';

const ItemIcon = styled(Icon)`
  display: flex;
  justify-content: center;
  align-items: center;
  margin-right: 8px;
`;

ItemIcon.displayName = 'ItemIcon';

const bulletValue = ({ bullet }) =>
  bullet &&
  css`
    &:before {
      content: '${bullet}';
    }
  `;

const ListItem = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;

  padding: 10px 16px;

  ${bulletValue};
`;

ListItem.displayName = 'ListItem';

class Item extends React.Component {
  static create = item => {
    if (typeof item === 'string') {
      return <Item content={item} key={item} />;
    }

    return <Item {...item} key={item.content} />;
  };

  _renderIcon = icon => {
    if (typeof icon === 'string') {
      return <ItemIcon name={icon} />;
    }

    return <ItemIcon {...icon} />;
  };

  render() {
    const { icon, children, content, bullet, ...rest } = this.props;

    return (
      <li {...rest}>
        <ListItem bullet={bullet}>
          {icon && this._renderIcon(icon)}
          {children || <Content content={content} />}
        </ListItem>
      </li>
    );
  }
}

Item.defaultProps = {
  icon: '',
  bullet: '',
  children: null,
  content: '',
};

Item.propTypes = {
  content: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      header: PropTypes.string,
      subheader: PropTypes.string,
    }),
  ]),
  icon: PropTypes.oneOfType([PropTypes.string, PropTypes.instanceOf(Object)]),
  bullet: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
};

Item.displayName = 'List.Item';

export default Item;
