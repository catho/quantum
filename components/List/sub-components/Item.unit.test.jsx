import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Item from './Item';
import Content from './Content';

const content = 'content';
const icon = 'info';
const header = 'header';
const subheader = 'subheader';

describe('<Item />', () => {
  it('Should match the snapshot', () => {
    const wrapper = mount(<Item />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('render an Item with icon', () => {
    const wrapper = mount(<Item icon={icon} />);

    expect(wrapper.find('ItemIcon')).toHaveLength(1);
  });

  it('render Content component', () => {
    const wrapper = mount(<Item content={content} />);

    expect(wrapper.render().text()).toBe(content);
    expect(wrapper.find(Content)).toHaveLength(1);
  });

  it('accept any children as a content', () => {
    const wrapper = mount(<Item>{content}</Item>);

    expect(wrapper.render().text()).toBe(content);
    expect(wrapper.find(Content)).toHaveLength(0);
  });

  it('render Content, Header and SubHeader components', () => {
    const wrapper = mount(<Item content={{ header, subheader }} />);

    expect(wrapper.find(Content)).toHaveLength(1);
    expect(wrapper.render().text()).toContain(header);
    expect(wrapper.render().text()).toContain(subheader);
  });
});
