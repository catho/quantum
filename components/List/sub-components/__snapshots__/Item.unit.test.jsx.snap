// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Item /> Should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <List.Item
    bullet=""
    content={null}
    icon=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ForwardRef
        bullet=""
      >
        
        <List.Content
          content={null}
          header=""
          subheader=""
        />
      </ForwardRef>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "bullet": "",
        "children": Array [
          "",
          <List.Content
            content={null}
            header=""
            subheader=""
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": null,
            "content": null,
            "header": "",
            "subheader": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Item__ListItem-culzzj-1",
          "isStatic": false,
          "rules": Array [
            "display:flex;flex-direction:row;align-items:center;padding:10px 16px;",
            [Function],
            ";",
          ],
        },
        "displayName": "ListItem",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Item__ListItem-culzzj-1",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
    "type": "li",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ForwardRef
          bullet=""
        >
          
          <List.Content
            content={null}
            header=""
            subheader=""
          />
        </ForwardRef>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "bullet": "",
          "children": Array [
            "",
            <List.Content
              content={null}
              header=""
              subheader=""
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          "",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": null,
              "content": null,
              "header": "",
              "subheader": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Item__ListItem-culzzj-1",
            "isStatic": false,
            "rules": Array [
              "display:flex;flex-direction:row;align-items:center;padding:10px 16px;",
              [Function],
              ";",
            ],
          },
          "displayName": "ListItem",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Item__ListItem-culzzj-1",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
      "type": "li",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
